### ISSUE TITLE
Divergent Gas Calculation for Calls in Replay Tests

### ISSUE THREAD

**Issue by DavePearce** [Consensys/besu-sequencer-plugins#953]
## Ranges

```rust
6110036-6110036
6110040-6110040
6110042-6110042
6110043-6110043
6110045-6110045
```

## Issue

_(see also #877)_

There remain situations where the gas limit calculated for e.g. a `CALL` in a replay test deviates from that reported in [LineaScan](https://lineascan.build/).  This can then lead to a `revert` of some kind (though doesn't always).  Example transactions where this occurs:

- [Block 6110045](https://lineascan.build/tx/0xe6dddb5a975fbb7b80a03669c789711069aba9372593837c8bab05938d8c4333#internal), tx `0xe6dddb5a975fbb...`.  Gas available for `call_0_1_1` is `221288` but should be `224044` (difference is `2756`).
- [Block 6110043](https://lineascan.build/tx/0x4ca3bfcb7db8cffd4c06c8247cb612cfecba3491fd6c116d988c4502620242b3#internal), tx `0x4ca3bfcb7db`.  Gas available for `call_0_1_1` is `45925` but should be `48681` (difference is `2756`).
- [Block 6110042](https://lineascan.build/tx/0x6def94778eb3b6b73f4597d4fdd9aa2af6c9e4a76fc1df6f5b0380fb2a222c63#internal) tx `0x6def94778`.  Gas available for `call_0_1_1_1` is `225924` but should be `228681` (difference is `2757`).
- [Block 6110036](https://lineascan.build/tx/0xb0b483fef9d76f4be0750ef82c220d4c1c06601ea30f56978e888ed5435dd851#internal) tx `0xb0b483fef9d`.  Gas available for `staticcall_0_1_1_1_1` is `641851` but should be `644608` (difference is `2757`).

Finally, there are also transactions where all internal transactions have the right amount of gas, but it still reverts.  For example, in [Block 6110040](https://lineascan.build/tx/0xe3e410ddcd6286a9e67ea75f82c654c81c49ec5d2696b7059885af7bf5b2316c#internal) tx `0xe3e410ddcd6286a9e` we see this.

## Reproduction

To see this divergence in gas, we need to instrument `ZkTracer` with something like this:

```
public void tracePrepareTransaction(WorldView worldView, Transaction transaction) {
    System.out.println("----");
    System.out.println("TX " + transaction.getHash());
    System.out.println("FROM " + transaction.getSender());
    System.out.println("TO " + transaction.getTo());
    ...
}

...

public void traceContextEnter(MessageFrame frame) {
  System.out.println(frame.getSenderAddress() + " => " + frame.getContractAddress() + " " +
      "(gas: " + frame.getRemainingGas() + ")");
  ...
}
```

Running this on the replay for e.g. block 6110045 gives this output:

```
----
TX 0xe6dddb5a975fbb7b80a03669c789711069aba9372593837c8bab05938d8c4333
FROM 0x41ee28ee05341e7fdddc8d433ba66054cd302ca1
TO Optional[0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75]
0x41ee28ee05341e7fdddc8d433ba66054cd302ca1 => 0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75 (gas: 318596)
0x41ee28ee05341e7fdddc8d433ba66054cd302ca1 => 0x477f34de55874729fd4f7e60b121338e7aa95145 (gas: 308758)
0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75 => 0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f (gas: 258494)
0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75 => 0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f (gas: 240298)
0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f => 0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75 (gas: 2300)
0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f => 0x477f34de55874729fd4f7e60b121338e7aa95145 (gas: 1909)
0x7e63a5f1a8f0b4d0934b2f2327daed3f6bb2ee75 => 0x9a2fed1871a32636cb5f7b2c837998bd1f1b9cb9 (gas: 221288)
----
TX 0xc8574d2e389a6474f06e635870bb1a3b2c5fe73f515297733545f305861d4428
FROM 0xea5a33eb17b8e00ca904ef090da0dd14b3dcc38b
TO Optional[0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7]
...
```

And we can see the gas `221288` in the last call of the first transaction.  Comparing with the [internal transactions](https://lineascan.build/tx/0xe6dddb5a975fbb7b80a03669c789711069aba9372593837c8bab05938d8c4333#internal) for that tx, you can see they all line up except that last one.

## Notes
* In my tests, I have eliminated side-effects from the constraints by commenting out all calls to the `hub` and `pin55`, etc.
* `2757` is `2800 - 2800/64`.  That means the issue is probably an unexpected `G_SRESET` charge (`2900`), suggesting something is up with an `SSTORE`.  Perhaps the value for some location stored by the block capturer was wrong somehow?

**Comment by DavePearce** [comment]
Calcluation for `CALL` in [Block 6110045](https://lineascan.build/tx/0xe6dddb5a975fbb7b80a03669c789711069aba9372593837c8bab05938d8c4333#internal), tx 0xe6dddb5a975fbb...: 

* `231563 - 100 - 9000 = 222463` (initial charge, assuming warm access) 
* `222463 - 222463/64 = 218988`
* `218988 + 2300 = 221288` (gas cap)

So, that would seem to be correct and suggests the gas on entry is wrong.

**Comment by DavePearce** [comment]
Ok, so I am seeing a problem with `ToyWorld`.  It has a bug related (presumably) to commiting of storage.  Yes, see #964 . This impacts [Block 6110040](https://lineascan.build/tx/0xe3e410ddcd6286a9e67ea75f82c654c81c49ec5d2696b7059885af7bf5b2316c#internal) tx `0xe3e410ddcd6286a9e`.  However, it does not resolve the other blocks listed at the beginning.

**Comment by DavePearce** [comment]
I believe this has been resolved as part of #1027 / #1048
